File: src-model/model/celestials/parts/Sphere.java
package model.celestials.parts;

import model.celestials.CelestialBody;
import model.type.DrawableResolution;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.jogamp.opengl.util.texture.Texture;

import engine.calculators.PhysicsCalculator;

/**
 * Object which stores the sphere parameters for {@link CelestialBody} entity.
 * 
 * TODO It should be upgraded by:<br>
 * - mass - json ignore, calculated from volume and density<br>
 * - density<br>
 */
public class Sphere {

    private Ring ring;
    private Clouds clouds;

    private String textureFile;

    private double radius;
    private double velocity;
    private double rotation;
    private double inclination;

    /* ========== PUBLIC ========== */
    @JsonCreator
    public Sphere(String textureFile, Clouds clouds, Ring ring, double radius, double velocity, double inclination) {

        this.ring = ring;
        this.clouds = clouds;

        this.textureFile = textureFile;

        this.radius = radius;
        this.velocity = velocity;
        this.inclination = inclination;

        this.rotation = 0;
        this.angularVelocity = PhysicsCalculator.angularVelocity(radius, velocity);
        this.resolution = DrawableResolution.VERY_FAR;
    }

    /* ========== PRIVATE ========== */
    @JsonIgnore
    private Texture texture;
    @JsonIgnore
    private double angularVelocity;
    @JsonIgnore
    private DrawableResolution resolution;
    @JsonIgnore
    private double headingRotation;

    /* ========== PROPERTIES ========== */
    public Ring getRing() {
        return ring;
    }

    public void setRing(Ring ring) {
        this.ring = ring;
    }

    public Clouds getClouds() {
        return clouds;
    }

    public void setClouds(Clouds clouds) {
        this.clouds = clouds;
    }

    public String getTextureFile() {
        return textureFile;
    }

    public void setTextureFile(String textureFile) {
        this.textureFile = textureFile;
    }

    public double getRadius() {
        return radius;
    }

    public void setRadius(double radius) {
        this.radius = radius;
    }

    public double getVelocity() {
        return velocity;
    }

    public void setVelocity(double velocity) {
        this.velocity = velocity;
    }

    public double getRotation() {
        return rotation;
    }

    public void setRotation(double rotation) {
        this.rotation = rotation;
    }

    public double getInclination() {
        return inclination;
    }

    public void setInclination(double inclination) {
        this.inclination = inclination;
    }

    public Texture getTexture() {
        return texture;
    }

    public void setTexture(Texture texture) {
        this.texture = texture;
    }

    public double getAngularVelocity() {
        return angularVelocity;
    }

    public void setAngularVelocity(double angularVelocity) {
        this.angularVelocity = angularVelocity;
    }

    public DrawableResolution getResolution() {
        return resolution;
    }

    public void setResolution(DrawableResolution resolution) {
        this.resolution = resolution;
    }

    public double getHeadingRotation() {
        return headingRotation;
    }

    public void setHeadingRotation(double headingRotation) {
        this.headingRotation = headingRotation;
    }
}

File: src-model/model/celestials/parts/Sphere.kt
package model.celestials.parts

import engine.calculators.PhysicsCalculator
import engine.components.drawers.DrawableSphere
import engine.math.AngularVelocity
import engine.math.Degree
import engine.math.Radius
import engine.math.Velocity
import engine.utils.TextureLoader
import model.celestials.CelestialBody
import model.interfaces.Drawable
import model.type.DrawableResolution
import javax.media.opengl.GL2

/**
 * Object which stores the sphere parameters for [CelestialBody] entity.
 *
 * TODO It should be upgraded by:<br></br>
 * - mass - json ignore, calculated from volume and density<br></br>
 * - density<br></br>
 */
class Sphere(
        var textureFile: String?,
        var radius: Radius,
        rotationVelocity: Velocity,
        var inclination: Degree
) : Drawable {
    var rotation: Degree = Degree.ZERO

    var angularVelocity: AngularVelocity = AngularVelocity(radius, rotationVelocity)
    var resolution: DrawableResolution = DrawableResolution.VERY_FAR
    var drawable: DrawableSphere? = null

    fun radius(): Radius {
        return radius
    }

    override fun draw(gl: GL2) {
        if (this.drawable == null) {
            this.drawable = DrawableSphere(this, TextureLoader.getTexture(gl, textureFile))
        }

        this.drawable?.draw(gl)
    }

    fun tick() {
        rotateSphere()
    }

    private fun rotateSphere() {
        rotation = (rotation + angularVelocity).normalize()
    }
}